src/OpenDiffix.Core/Analyzer.fs

```diff
@@ -309,73 +309,63 @@ let private bucketExpand aidColumnsExpression =
+
+  let selectQuery =
+    if noGrouping && noAggregation then
+      // Non-aggregating query; group implicitly and expand.
+      { selectQuery with
+          TargetList =
+            {
+              Expression = bucketExpand aidColumnsExpression
+              Alias = ""
+              Tag = JunkTargetEntry
+            }
+            :: selectQuery.TargetList
+          GroupBy = selectedExpressions
+      }
+    else
+      selectQuery
 
-  let doesGrouping = List.isEmpty selectQuery.GroupBy |> not
-  let doesAggregation = selectedExpressions |> List.forall Expression.isScalar |> not
-  let onlyConstantsSelected = List.isEmpty nonConstantExpressions
-
-  if not doesGrouping && not doesAggregation then
-    // Non-grouping, non-aggregate query; group implicitly and expand
-    let having =
-      if not onlyConstantsSelected then
-        Expression.makeAnd lowCountFilter selectQuery.Having
-      else
-        // All selected expressions are constants; no low-count filter in line with global anonymized count.
-        selectQuery.Having
-
-    { selectQuery with
-        TargetList =
-          {
-            Expression = bucketExpand aidColumnsExpression
-            Alias = ""
-            Tag = JunkTargetEntry
-          }
-          :: selectQuery.TargetList
-        // No need to group implicitly by constants, which are invalid bucket definitions anyway.
-        GroupBy = nonConstantExpressions
-        Having = having
-    }
-  else if not doesGrouping && doesAggregation then
-    // Non-grouping aggregate query; do nothing.
-    selectQuery
-  else
-    // Grouping query; add LCF to HAVING
-    { selectQuery with
-        Having = Expression.makeAnd lowCountFilter selectQuery.Having
-    }
+  // Ignore constant expressions during grouped aggregation.
+  let groupBy = List.filter (isConstant >> not) selectQuery.GroupBy
 
-let private rewriteQuery anonParams (selectQuery: SelectQuery) =
+  let having =
+    if List.isEmpty groupBy then
+      // No need to apply the LCF to the global bucket.
+      selectQuery.Having
+    else
+      // Apply the LCF to the output buckets.
+      [ aidColumnsExpression ]
+      |> Expression.makeAggregate DiffixLowCount
+      |> Expression.makeNot
+      |> Expression.makeAnd selectQuery.Having
+
+  { selectQuery with GroupBy = groupBy; Having = having }
+
+let private compileQuery anonParams (selectQuery: SelectQuery) =
   let rangeColumns = collectRangeColumns anonParams selectQuery.From
   let aidColumnsExpression = makeAidColumnsExpression rangeColumns
 
-  let isAnonymizing =
-    anonParams.AccessLevel <> Direct
-    && aidColumnsExpression |> Expression.unwrapListExpr |> List.isEmpty |> not
+  let directAccess =
+    anonParams.AccessLevel = Direct
+    || aidColumnsExpression |> Expression.unwrapListExpr |> List.isEmpty
 
-  QueryValidator.validateQuery isAnonymizing anonParams.AccessLevel selectQuery
+  QueryValidator.validateStandardQuery selectQuery
```

Comment:
not about your change, but I'm also finding the term `standard query` confusing - even despite already knowing what that means. Even more so, that the validation involved is about "non-standard" (in layman terms) concepts like `DiffixLowCount` aggregators.

How about we rename this globally to `direct query` (it's probably spread across the repo, but not very frequent, should be manageable)? Then it lines up with `direct access` nicely, and better reflects what is really happening (if we're using Diffix, we're using it directly).

Reply:
I'm not convinced that fits well with the intent of this function.
Calling this a `direct query` implies that it isn't an `anonymizing query`. This function should execute common SQL validations that apply to both direct and anonymizing queries. Maybe we should have 3 entry points here: standard SQL checks, direct access checks and anonymizing checks? I will leave it up to you to decide.

Reply:
yeah, same issue with `standard query`, we run this even if the query isn't standard. I'll propose something

Reply:
In my head, this somewhat makes sense: the standard SQL validations also apply to anonymizing queries.

Reply:
To me, the confusion is in the gist of the name `standard query`, in some of the context it appears in here. Also in the fact that `validateStandardQuery` doesn't do checks relevant to anonymizing query. Realizing the latter made it pretty straightforward to untangle, let me show my refactor in a minute...